src/serial_open.cpp

```diff
@@ -0,0 +1,201 @@
+#include <cpp_core/interface/serial_open.h>
+#include <cpp_core/status_codes.h>
+
+#include <cerrno>
+#include <fcntl.h>
+#include <sys/ioctl.h>
+#include <system_error>
+#include <termios.h>
+#include <unistd.h>
+
+#ifndef TCGETS2
+#define TCGETS2 0x802C542A
+#define TCSETS2 0x402C542B
+#define BOTHER 0x010000
+#endif
+
+// NOLINTBEGIN
+// C Structure is defined by the kernel, so we cannot change it.
+struct termios2
+{
+    tcflag_t c_iflag;
+    tcflag_t c_oflag;
+    tcflag_t c_cflag;
+    tcflag_t c_lflag;
+    cc_t c_line;
+    cc_t c_cc[19];
+    speed_t c_ispeed;
+    speed_t c_ospeed;
+};
+// NOLINTEND
+
+extern "C"
+{
+    // NOLINTNEXTLINE(readability-function-cognitive-complexity)
+    MODULE_API auto serialOpen(void *port, int baudrate, int data_bits, int parity, int stop_bits,
+                               ErrorCallbackT error_callback) -> intptr_t
+    {
+        if (port == nullptr)
+        {
+            if (error_callback != nullptr)
+            {
+                error_callback(static_cast<int>(cpp_core::StatusCodes::kNotFoundError), "Port parameter is nullptr");
+            }
+            return static_cast<intptr_t>(cpp_core::StatusCodes::kNotFoundError);
+        }
+
+        if (baudrate < 300)
+        {
+            if (error_callback != nullptr)
+            {
+                error_callback(static_cast<int>(cpp_core::StatusCodes::kSetStateError),
+                               "Invalid baudrate: must be >= 300");
+            }
+            return static_cast<intptr_t>(cpp_core::StatusCodes::kSetStateError);
+        }
+
+        if (data_bits < 5 || data_bits > 8)
+        {
+            if (error_callback != nullptr)
+            {
+                error_callback(static_cast<int>(cpp_core::StatusCodes::kSetStateError),
+                               "Invalid data bits: must be 5-8");
+            }
+            return static_cast<intptr_t>(cpp_core::StatusCodes::kSetStateError);
+        }
+
+        const char *port_path = static_cast<const char *>(port);
+
+        const int fd = open(port_path, O_RDWR | O_NOCTTY | O_NONBLOCK);
```

Comment:
I am assuming that `fd` stands for `file_descriptor`? Pleas use variable names that are clearly.

```suggestion
        const int handle = open(port_path, O_RDWR | O_NOCTTY | O_NONBLOCK);
```

Reply:
fd is the kernel default for file descriptors, since its not just a handle